input-events/input-events-typing.html

```diff
@@ -1,82 +1,294 @@
 <!DOCTYPE html>
 <meta charset="utf-8">
-<title>InputEvent have correct data/order when typing on textarea and contenteditable</title>
+<title>Input Event typing tests</title>
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <script src="/resources/testdriver.js"></script>
 <script src="/resources/testdriver-vendor.js"></script>
 <script src="/resources/testdriver-actions.js"></script>
-<p>To manually run this test, please follow the steps below:<br/>
-1. Focus the first box and press key 'a' and then Shift-'b'.<br/>
-2. Focus the second box and press key 'c' and then Shift-'d'.<br/>
-<br/>
-If a "PASS" result appears the test passes, otherwise it fails</p>
-<textarea id='plain'></textarea>
-<div id='rich' style='border-style: solid;' contenteditable><br></div>
+<div id="rich" contenteditable></div>
+<textarea id="plain"></textarea>
 <script>
-async_test(t => {
+let inputEventsLog = [];
+const rich = document.getElementById('rich');
+const plain = document.getElementById('plain');
+
+function log(event) {
+    inputEventsLog.push({
+        type: event.type,
+        inputType: event.inputType,
+        data: event.data,
+        isComposing: event.isComposing,
+        cancelable: event.cancelable,
+        state: encodeURI(rich.innerHTML)
+    });
+}
+
+function resetRich() {
+    inputEventsLog = [];
+    rich.innerHTML = '';
+}
+
+function setRichContent(content = '') {
+    rich.innerHTML = content;
+}
+
+rich.addEventListener('beforeinput', event => log(event));
+rich.addEventListener('input', event => log(event));
+
+promise_test(async function() {
+    this.add_cleanup(resetRich);
+    rich.focus();
+    const message = 'Hello';
+    await test_driver.send_keys(rich, message);
+    // 10 events (5 beforeinput + 5 input events)
+    assert_equals(inputEventsLog.length, 10);
+    for (let i = 0; i < inputEventsLog.length; i += 2) {
```

Comment:
Can we create a function to convert `inputEventsLog` to an array of event object pair (`[[(beforeinput), (input)]]`) so that we can just `for (const [beforeInputEvent, inputEvent] of ...)` here? It's slightly confusing to have `i + 1` and `i / 2` here.

Reply:
Feel free to refactor the test. For me it looks the most readable the way it currently is, but we are all different and I'm OK with adjusting the test so that it fits in better with other tests.

Not though that for the first test we need to note down what is in the editor at the time of the event (state).

Reply:
@saschanaz If you do refactor - notice that one of the things this tests for is whether there actually is a beforeinput-input event pair or not. In the past there have been browsers that just emitted an input event and two of these tests currently fail in Chrome because it emits a beforeinput event but no input event. 

So that converter function should probably detect such cases and then let the test fail in some way that makes it clear that that is what happened.

Reply:
Ok, let's do this as a followup (if we do it 🙃).